<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Data Exploration | Comparative genomics using R</title>
  <meta name="description" content="Chapter 2 Data Exploration | Comparative genomics using R" />
  <meta name="generator" content="bookdown 0.12 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Data Exploration | Comparative genomics using R" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://github.com/StefanoBertoUTSW/PEB_BOOK" />
  <meta property="og:image" content="https://github.com/StefanoBertoUTSW/PEB_BOOKimages/cover.jpg" />
  <meta property="og:description" content="Chapter 2 Data Exploration | Comparative genomics using R" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Data Exploration | Comparative genomics using R" />
  
  <meta name="twitter:description" content="Chapter 2 Data Exploration | Comparative genomics using R" />
  <meta name="twitter:image" content="https://github.com/StefanoBertoUTSW/PEB_BOOKimages/cover.jpg" />

<meta name="author" content="Stefano Berto, PhD" />


<meta name="date" content="2020-03-01" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introPipes.html">
<link rel="next" href="DiffExpAnalysis.html">
<script src="libs/jquery/jquery.min.js"></script>
<link href="libs/gitbook/css/style.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-fontsettings.css" rel="stylesheet" />







<script src="libs/htmlwidgets/htmlwidgets.js"></script>
<link href="libs/datatables-css/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding/datatables.js"></script>
<link href="libs/dt-core/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk/js/crosstalk.min.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Comparative Genomics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Introduction</a><ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#what-will-you-get-out-of-this"><i class="fa fa-check"></i>What will you get out of this?</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#structure-of-the-course"><i class="fa fa-check"></i>Structure of the course</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#packages-needed-to-run-the-course-code"><i class="fa fa-check"></i>Packages needed to run the course code</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introPipes.html"><a href="introPipes.html"><i class="fa fa-check"></i><b>1</b> Introduction to RNA-seq pipes</a><ul>
<li class="chapter" data-level="1.1" data-path="introPipes.html"><a href="introPipes.html#sequencing-machines"><i class="fa fa-check"></i><b>1.1</b> Sequencing Machines</a></li>
<li class="chapter" data-level="1.2" data-path="introPipes.html"><a href="introPipes.html#fastq"><i class="fa fa-check"></i><b>1.2</b> FASTQ</a></li>
<li class="chapter" data-level="1.3" data-path="introPipes.html"><a href="introPipes.html#quality-check"><i class="fa fa-check"></i><b>1.3</b> Quality Check</a></li>
<li class="chapter" data-level="1.4" data-path="introPipes.html"><a href="introPipes.html#quality-trimming"><i class="fa fa-check"></i><b>1.4</b> Quality Trimming</a></li>
<li class="chapter" data-level="1.5" data-path="introPipes.html"><a href="introPipes.html#alignment"><i class="fa fa-check"></i><b>1.5</b> Alignment</a></li>
<li class="chapter" data-level="1.6" data-path="introPipes.html"><a href="introPipes.html#statistics-and-filtering"><i class="fa fa-check"></i><b>1.6</b> Statistics and Filtering</a></li>
<li class="chapter" data-level="1.7" data-path="introPipes.html"><a href="introPipes.html#lift-over"><i class="fa fa-check"></i><b>1.7</b> Lift over</a></li>
<li class="chapter" data-level="1.8" data-path="introPipes.html"><a href="introPipes.html#quantification"><i class="fa fa-check"></i><b>1.8</b> Quantification</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="DataExploration.html"><a href="DataExploration.html"><i class="fa fa-check"></i><b>2</b> Data Exploration</a><ul>
<li class="chapter" data-level="2.0.1" data-path="DataExploration.html"><a href="DataExploration.html#data-loading"><i class="fa fa-check"></i><b>2.0.1</b> Data loading</a></li>
<li class="chapter" data-level="2.0.2" data-path="DataExploration.html"><a href="DataExploration.html#demographic-screening"><i class="fa fa-check"></i><b>2.0.2</b> Demographic screening</a></li>
<li class="chapter" data-level="2.0.3" data-path="DataExploration.html"><a href="DataExploration.html#data-normalization"><i class="fa fa-check"></i><b>2.0.3</b> Data normalization</a></li>
<li class="chapter" data-level="2.0.4" data-path="DataExploration.html"><a href="DataExploration.html#initial-data-exploration"><i class="fa fa-check"></i><b>2.0.4</b> Initial Data Exploration</a></li>
<li class="chapter" data-level="2.0.5" data-path="DataExploration.html"><a href="DataExploration.html#variance-explained-by-covariates"><i class="fa fa-check"></i><b>2.0.5</b> Variance explained by covariates</a></li>
<li class="chapter" data-level="2.0.6" data-path="DataExploration.html"><a href="DataExploration.html#save-the-data"><i class="fa fa-check"></i><b>2.0.6</b> Save the data</a></li>
<li class="chapter" data-level="2.0.7" data-path="DataExploration.html"><a href="DataExploration.html#exercise-for-data-exploration-chapter"><i class="fa fa-check"></i><b>2.0.7</b> Exercise for Data Exploration chapter</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="DiffExpAnalysis.html"><a href="DiffExpAnalysis.html"><i class="fa fa-check"></i><b>3</b> Differential Expression Analysis</a><ul>
<li class="chapter" data-level="3.1" data-path="DiffExpAnalysis.html"><a href="DiffExpAnalysis.html#data-loading-1"><i class="fa fa-check"></i><b>3.1</b> Data loading</a></li>
<li class="chapter" data-level="3.2" data-path="DiffExpAnalysis.html"><a href="DiffExpAnalysis.html#dge-based-on-linear-model"><i class="fa fa-check"></i><b>3.2</b> DGE based on linear model</a></li>
<li class="chapter" data-level="3.3" data-path="DiffExpAnalysis.html"><a href="DiffExpAnalysis.html#species-specific-dge"><i class="fa fa-check"></i><b>3.3</b> Species Specific DGE</a></li>
<li class="chapter" data-level="3.4" data-path="DiffExpAnalysis.html"><a href="DiffExpAnalysis.html#surrogate-variables"><i class="fa fa-check"></i><b>3.4</b> Surrogate Variables</a></li>
<li class="chapter" data-level="3.5" data-path="DiffExpAnalysis.html"><a href="DiffExpAnalysis.html#balance-the-gene-expression-for-covariates"><i class="fa fa-check"></i><b>3.5</b> Balance the gene expression for covariates</a></li>
<li class="chapter" data-level="3.6" data-path="DiffExpAnalysis.html"><a href="DiffExpAnalysis.html#dge-visualizations"><i class="fa fa-check"></i><b>3.6</b> DGE visualizations</a></li>
<li class="chapter" data-level="3.7" data-path="DiffExpAnalysis.html"><a href="DiffExpAnalysis.html#functional-enrichment"><i class="fa fa-check"></i><b>3.7</b> Functional Enrichment</a></li>
<li class="chapter" data-level="3.8" data-path="DiffExpAnalysis.html"><a href="DiffExpAnalysis.html#dge-based-on-deseq2"><i class="fa fa-check"></i><b>3.8</b> DGE based on DESeq2</a></li>
<li class="chapter" data-level="3.9" data-path="DiffExpAnalysis.html"><a href="DiffExpAnalysis.html#exercise-for-differential-expression-chapter"><i class="fa fa-check"></i><b>3.9</b> Exercise for Differential Expression Chapter</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Comparative genomics using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="DataExploration" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Data Exploration</h1>
<p>The starting data consist in RNA-seq count matrix after quantification (exp), a demographic table (demo) and a gene length table (width). The count table is a integer matrix without any normalizations. Each rows correspond to a gene, each column represent a different sample from three species (human, chimpanzee and rhesus macaque). The genes are protein-coding and orthologous between the three species.</p>
<div id="data-loading" class="section level3">
<h3><span class="header-section-number">2.0.1</span> Data loading</h3>
<p><strong>So let’s start!</strong></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co"># First load the libraries we will use in this section</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(sva))</a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(DESeq2))</a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(limma))</a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(tidyverse))</a>
<a class="sourceLine" id="cb10-6" data-line-number="6"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(openxlsx))</a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(ggplot2))</a>
<a class="sourceLine" id="cb10-8" data-line-number="8"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(ggpubr))</a>
<a class="sourceLine" id="cb10-9" data-line-number="9"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(pheatmap))</a>
<a class="sourceLine" id="cb10-10" data-line-number="10"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(here))</a>
<a class="sourceLine" id="cb10-11" data-line-number="11"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(future.apply))</a>
<a class="sourceLine" id="cb10-12" data-line-number="12"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(broom))</a>
<a class="sourceLine" id="cb10-13" data-line-number="13"></a>
<a class="sourceLine" id="cb10-14" data-line-number="14"><span class="co"># Clean the env</span></a>
<a class="sourceLine" id="cb10-15" data-line-number="15"><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</a>
<a class="sourceLine" id="cb10-16" data-line-number="16"></a>
<a class="sourceLine" id="cb10-17" data-line-number="17"><span class="co"># the data is stored under the subdirectory &#39;data&#39;</span></a>
<a class="sourceLine" id="cb10-18" data-line-number="18"><span class="kw">list.files</span>()</a>
<a class="sourceLine" id="cb10-19" data-line-number="19"><span class="co">##  [1] &quot;_book&quot;                  &quot;_bookdown_files&quot;       </span></a>
<a class="sourceLine" id="cb10-20" data-line-number="20"><span class="co">##  [3] &quot;_bookdown.yml&quot;          &quot;_build.sh&quot;             </span></a>
<a class="sourceLine" id="cb10-21" data-line-number="21"><span class="co">##  [5] &quot;_deploy.sh&quot;             &quot;_output.yml&quot;           </span></a>
<a class="sourceLine" id="cb10-22" data-line-number="22"><span class="co">##  [7] &quot;01-introPipes.Rmd&quot;      &quot;02-DataExploration.Rmd&quot;</span></a>
<a class="sourceLine" id="cb10-23" data-line-number="23"><span class="co">##  [9] &quot;03-DiffExpAnalysis.Rmd&quot; &quot;addson&quot;                </span></a>
<a class="sourceLine" id="cb10-24" data-line-number="24"><span class="co">## [11] &quot;book.bib&quot;               &quot;bookdown-demo_cache&quot;   </span></a>
<a class="sourceLine" id="cb10-25" data-line-number="25"><span class="co">## [13] &quot;bookdown-demo_files&quot;    &quot;bookdown-demo.Rmd&quot;     </span></a>
<a class="sourceLine" id="cb10-26" data-line-number="26"><span class="co">## [15] &quot;bookdown-demo.Rproj&quot;    &quot;DESCRIPTION&quot;           </span></a>
<a class="sourceLine" id="cb10-27" data-line-number="27"><span class="co">## [17] &quot;Dockerfile&quot;             &quot;images&quot;                </span></a>
<a class="sourceLine" id="cb10-28" data-line-number="28"><span class="co">## [19] &quot;index.Rmd&quot;              &quot;LICENSE&quot;               </span></a>
<a class="sourceLine" id="cb10-29" data-line-number="29"><span class="co">## [21] &quot;now.json&quot;               &quot;packages.bib&quot;          </span></a>
<a class="sourceLine" id="cb10-30" data-line-number="30"><span class="co">## [23] &quot;peb_data&quot;               &quot;preamble.tex&quot;          </span></a>
<a class="sourceLine" id="cb10-31" data-line-number="31"><span class="co">## [25] &quot;README.md&quot;              &quot;style.css&quot;             </span></a>
<a class="sourceLine" id="cb10-32" data-line-number="32"><span class="co">## [27] &quot;toc.css&quot;</span></a>
<a class="sourceLine" id="cb10-33" data-line-number="33"></a>
<a class="sourceLine" id="cb10-34" data-line-number="34"><span class="co"># load the data you need</span></a>
<a class="sourceLine" id="cb10-35" data-line-number="35"><span class="kw">load</span>(<span class="kw">here</span>(<span class="st">&quot;peb_data&quot;</span>, <span class="st">&quot;PEB_2020.RData&quot;</span>))</a>
<a class="sourceLine" id="cb10-36" data-line-number="36"></a>
<a class="sourceLine" id="cb10-37" data-line-number="37"><span class="co"># Check what data you loaded</span></a>
<a class="sourceLine" id="cb10-38" data-line-number="38"><span class="kw">ls</span>()</a>
<a class="sourceLine" id="cb10-39" data-line-number="39"><span class="co">## [1] &quot;demo&quot;  &quot;exp&quot;   &quot;width&quot;</span></a></code></pre></div>
<p>The <strong>exp</strong> is the data frame containing raw counts.</p>
<p>The <strong>demo</strong> is the demographic data with some biological and technical covariates</p>
<p>The <strong>width</strong> is the length of the the genes for data normalizations.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># Check the dimension of exp</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw">dim</span>(exp)</a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="co">## [1] 13291    30</span></a>
<a class="sourceLine" id="cb11-4" data-line-number="4"></a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="co"># Let&#39;s have a look to the data distribution of the</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="co"># expression</span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="kw">summary</span>(exp)</a>
<a class="sourceLine" id="cb11-8" data-line-number="8"><span class="co">##      Hsap_1           Hsap_2           Hsap_3      </span></a>
<a class="sourceLine" id="cb11-9" data-line-number="9"><span class="co">##  Min.   :     0   Min.   :     0   Min.   :     0  </span></a>
<a class="sourceLine" id="cb11-10" data-line-number="10"><span class="co">##  1st Qu.:    28   1st Qu.:    13   1st Qu.:    14  </span></a>
<a class="sourceLine" id="cb11-11" data-line-number="11"><span class="co">##  Median :   464   Median :   156   Median :   195  </span></a>
<a class="sourceLine" id="cb11-12" data-line-number="12"><span class="co">##  Mean   :  3415   Mean   :  1209   Mean   :  1427  </span></a>
<a class="sourceLine" id="cb11-13" data-line-number="13"><span class="co">##  3rd Qu.:  2256   3rd Qu.:   787   3rd Qu.:   974  </span></a>
<a class="sourceLine" id="cb11-14" data-line-number="14"><span class="co">##  Max.   :755442   Max.   :252898   Max.   :266769  </span></a>
<a class="sourceLine" id="cb11-15" data-line-number="15"><span class="co">##      Hsap_4           Hsap_5           Hsap_6      </span></a>
<a class="sourceLine" id="cb11-16" data-line-number="16"><span class="co">##  Min.   :     0   Min.   :     0   Min.   :     0  </span></a>
<a class="sourceLine" id="cb11-17" data-line-number="17"><span class="co">##  1st Qu.:     9   1st Qu.:    13   1st Qu.:    12  </span></a>
<a class="sourceLine" id="cb11-18" data-line-number="18"><span class="co">##  Median :   110   Median :   163   Median :   176  </span></a>
<a class="sourceLine" id="cb11-19" data-line-number="19"><span class="co">##  Mean   :   977   Mean   :  1419   Mean   :  1217  </span></a>
<a class="sourceLine" id="cb11-20" data-line-number="20"><span class="co">##  3rd Qu.:   578   3rd Qu.:   882   3rd Qu.:   823  </span></a>
<a class="sourceLine" id="cb11-21" data-line-number="21"><span class="co">##  Max.   :208039   Max.   :310497   Max.   :235453  </span></a>
<a class="sourceLine" id="cb11-22" data-line-number="22"><span class="co">##      Hsap_7           Hsap_8           Hsap_9      </span></a>
<a class="sourceLine" id="cb11-23" data-line-number="23"><span class="co">##  Min.   :     0   Min.   :     0   Min.   :     0  </span></a>
<a class="sourceLine" id="cb11-24" data-line-number="24"><span class="co">##  1st Qu.:    19   1st Qu.:    14   1st Qu.:     9  </span></a>
<a class="sourceLine" id="cb11-25" data-line-number="25"><span class="co">##  Median :   238   Median :   142   Median :   132  </span></a>
<a class="sourceLine" id="cb11-26" data-line-number="26"><span class="co">##  Mean   :  1150   Mean   :  1080   Mean   :  1212  </span></a>
<a class="sourceLine" id="cb11-27" data-line-number="27"><span class="co">##  3rd Qu.:   990   3rd Qu.:   642   3rd Qu.:   704  </span></a>
<a class="sourceLine" id="cb11-28" data-line-number="28"><span class="co">##  Max.   :156780   Max.   :208555   Max.   :306040  </span></a>
<a class="sourceLine" id="cb11-29" data-line-number="29"><span class="co">##     Hsap_10          PanTro_1        PanTro_2     </span></a>
<a class="sourceLine" id="cb11-30" data-line-number="30"><span class="co">##  Min.   :     0   Min.   :    0   Min.   :     0  </span></a>
<a class="sourceLine" id="cb11-31" data-line-number="31"><span class="co">##  1st Qu.:     9   1st Qu.:    8   1st Qu.:    13  </span></a>
<a class="sourceLine" id="cb11-32" data-line-number="32"><span class="co">##  Median :   146   Median :   72   Median :   188  </span></a>
<a class="sourceLine" id="cb11-33" data-line-number="33"><span class="co">##  Mean   :  1312   Mean   :  363   Mean   :  1942  </span></a>
<a class="sourceLine" id="cb11-34" data-line-number="34"><span class="co">##  3rd Qu.:   774   3rd Qu.:  321   3rd Qu.:  1048  </span></a>
<a class="sourceLine" id="cb11-35" data-line-number="35"><span class="co">##  Max.   :291565   Max.   :42193   Max.   :492501  </span></a>
<a class="sourceLine" id="cb11-36" data-line-number="36"><span class="co">##     PanTro_3        PanTro_4         PanTro_5     </span></a>
<a class="sourceLine" id="cb11-37" data-line-number="37"><span class="co">##  Min.   :    0   Min.   :     0   Min.   :     0  </span></a>
<a class="sourceLine" id="cb11-38" data-line-number="38"><span class="co">##  1st Qu.:    0   1st Qu.:     9   1st Qu.:     7  </span></a>
<a class="sourceLine" id="cb11-39" data-line-number="39"><span class="co">##  Median :   50   Median :   139   Median :   113  </span></a>
<a class="sourceLine" id="cb11-40" data-line-number="40"><span class="co">##  Mean   :  312   Mean   :  1125   Mean   :  1218  </span></a>
<a class="sourceLine" id="cb11-41" data-line-number="41"><span class="co">##  3rd Qu.:  235   3rd Qu.:   697   3rd Qu.:   670  </span></a>
<a class="sourceLine" id="cb11-42" data-line-number="42"><span class="co">##  Max.   :60608   Max.   :257678   Max.   :214389  </span></a>
<a class="sourceLine" id="cb11-43" data-line-number="43"><span class="co">##     PanTro_6         PanTro_7        PanTro_8     </span></a>
<a class="sourceLine" id="cb11-44" data-line-number="44"><span class="co">##  Min.   :     0   Min.   :    0   Min.   :     0  </span></a>
<a class="sourceLine" id="cb11-45" data-line-number="45"><span class="co">##  1st Qu.:     6   1st Qu.:   24   1st Qu.:    12  </span></a>
<a class="sourceLine" id="cb11-46" data-line-number="46"><span class="co">##  Median :    69   Median :  268   Median :   173  </span></a>
<a class="sourceLine" id="cb11-47" data-line-number="47"><span class="co">##  Mean   :   575   Mean   : 1067   Mean   :  1447  </span></a>
<a class="sourceLine" id="cb11-48" data-line-number="48"><span class="co">##  3rd Qu.:   345   3rd Qu.: 1000   3rd Qu.:   910  </span></a>
<a class="sourceLine" id="cb11-49" data-line-number="49"><span class="co">##  Max.   :170033   Max.   :92973   Max.   :345260  </span></a>
<a class="sourceLine" id="cb11-50" data-line-number="50"><span class="co">##     PanTro_9        PanTro_10         RheMac_1     </span></a>
<a class="sourceLine" id="cb11-51" data-line-number="51"><span class="co">##  Min.   :     0   Min.   :     0   Min.   :     0  </span></a>
<a class="sourceLine" id="cb11-52" data-line-number="52"><span class="co">##  1st Qu.:    12   1st Qu.:    15   1st Qu.:     9  </span></a>
<a class="sourceLine" id="cb11-53" data-line-number="53"><span class="co">##  Median :   165   Median :   169   Median :   136  </span></a>
<a class="sourceLine" id="cb11-54" data-line-number="54"><span class="co">##  Mean   :   995   Mean   :   936   Mean   :  1053  </span></a>
<a class="sourceLine" id="cb11-55" data-line-number="55"><span class="co">##  3rd Qu.:   695   3rd Qu.:   727   3rd Qu.:   678  </span></a>
<a class="sourceLine" id="cb11-56" data-line-number="56"><span class="co">##  Max.   :155820   Max.   :139103   Max.   :210549  </span></a>
<a class="sourceLine" id="cb11-57" data-line-number="57"><span class="co">##     RheMac_2        RheMac_3         RheMac_4    </span></a>
<a class="sourceLine" id="cb11-58" data-line-number="58"><span class="co">##  Min.   :    0   Min.   :     0   Min.   :    0  </span></a>
<a class="sourceLine" id="cb11-59" data-line-number="59"><span class="co">##  1st Qu.:   13   1st Qu.:     5   1st Qu.:    6  </span></a>
<a class="sourceLine" id="cb11-60" data-line-number="60"><span class="co">##  Median :  156   Median :    76   Median :   98  </span></a>
<a class="sourceLine" id="cb11-61" data-line-number="61"><span class="co">##  Mean   :  713   Mean   :   803   Mean   :  538  </span></a>
<a class="sourceLine" id="cb11-62" data-line-number="62"><span class="co">##  3rd Qu.:  610   3rd Qu.:   432   3rd Qu.:  419  </span></a>
<a class="sourceLine" id="cb11-63" data-line-number="63"><span class="co">##  Max.   :85555   Max.   :188456   Max.   :87970  </span></a>
<a class="sourceLine" id="cb11-64" data-line-number="64"><span class="co">##     RheMac_5        RheMac_6         RheMac_7    </span></a>
<a class="sourceLine" id="cb11-65" data-line-number="65"><span class="co">##  Min.   :    0   Min.   :     0   Min.   :    0  </span></a>
<a class="sourceLine" id="cb11-66" data-line-number="66"><span class="co">##  1st Qu.:    9   1st Qu.:     7   1st Qu.:   12  </span></a>
<a class="sourceLine" id="cb11-67" data-line-number="67"><span class="co">##  Median :   98   Median :   103   Median :  148  </span></a>
<a class="sourceLine" id="cb11-68" data-line-number="68"><span class="co">##  Mean   :  492   Mean   :   919   Mean   :  706  </span></a>
<a class="sourceLine" id="cb11-69" data-line-number="69"><span class="co">##  3rd Qu.:  399   3rd Qu.:   522   3rd Qu.:  590  </span></a>
<a class="sourceLine" id="cb11-70" data-line-number="70"><span class="co">##  Max.   :52623   Max.   :199397   Max.   :64114  </span></a>
<a class="sourceLine" id="cb11-71" data-line-number="71"><span class="co">##     RheMac_8         RheMac_9       RheMac_10     </span></a>
<a class="sourceLine" id="cb11-72" data-line-number="72"><span class="co">##  Min.   :     0   Min.   :    0   Min.   :     0  </span></a>
<a class="sourceLine" id="cb11-73" data-line-number="73"><span class="co">##  1st Qu.:    14   1st Qu.:    8   1st Qu.:     4  </span></a>
<a class="sourceLine" id="cb11-74" data-line-number="74"><span class="co">##  Median :   177   Median :   78   Median :    65  </span></a>
<a class="sourceLine" id="cb11-75" data-line-number="75"><span class="co">##  Mean   :  1137   Mean   :  429   Mean   :   697  </span></a>
<a class="sourceLine" id="cb11-76" data-line-number="76"><span class="co">##  3rd Qu.:   750   3rd Qu.:  334   3rd Qu.:   365  </span></a>
<a class="sourceLine" id="cb11-77" data-line-number="77"><span class="co">##  Max.   :268860   Max.   :46577   Max.   :182266</span></a>
<a class="sourceLine" id="cb11-78" data-line-number="78"></a>
<a class="sourceLine" id="cb11-79" data-line-number="79"><span class="kw">boxplot</span>(exp)</a></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/checkData-1.png" width="65%" style="display: block; margin: auto;" /></p>
</div>
<div id="demographic-screening" class="section level3">
<h3><span class="header-section-number">2.0.2</span> Demographic screening</h3>
<p>Demographics are important to understand what type of data you are dealing with.
There are several factors that can influence expression: some are categorical variables (e.g. Sex, Hemisphere), others are continous variables (e.g. Age).
These typically are important covariates to take into account into the analysis.
In principle, RNA-seq is a snapshot of the gene expression in a specific moment of time from a specific tissue (bulk RNA-seq) or cell-type (single cell RNA-seq).</p>
<p>Different factors can play a role in explaining partially the variance of the gene expression you see.</p>
<p>For instance, gender can slightly differentiate gene expression therefore we presume it will have a minimal impact on gene expression variance. On the other hand variables like PMI (Post-mortem interval), RIN (rna integrity number), Brain Bank (the institute where you collect the data), Batch (sequencing days and hands), and Age (the age of the individual who donated the tissue) might have a sever impact on the gene expression variance.</p>
<p>These variables must be taken into account when RNA-seq (but also other genomics) is analyzed.</p>
<p>Now let’s see what we have in the demographic</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw">head</span>(demo)</a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="co">##        Species Sex Age RIN Hemisphere PMI</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="co">## Hsap_1    Hsap   M  42 9.8          L  13</span></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="co">## Hsap_2    Hsap   M  39 6.4          L  11</span></a>
<a class="sourceLine" id="cb12-5" data-line-number="5"><span class="co">## Hsap_3    Hsap   F  25 8.8          R  18</span></a>
<a class="sourceLine" id="cb12-6" data-line-number="6"><span class="co">## Hsap_4    Hsap   F  21 7.9          L   6</span></a>
<a class="sourceLine" id="cb12-7" data-line-number="7"><span class="co">## Hsap_5    Hsap   M  45 6.6          L  16</span></a>
<a class="sourceLine" id="cb12-8" data-line-number="8"><span class="co">## Hsap_6    Hsap   M  25 5.8          R   5</span></a>
<a class="sourceLine" id="cb12-9" data-line-number="9"></a>
<a class="sourceLine" id="cb12-10" data-line-number="10"><span class="kw">str</span>(demo)</a>
<a class="sourceLine" id="cb12-11" data-line-number="11"><span class="co">## &#39;data.frame&#39;:    30 obs. of  6 variables:</span></a>
<a class="sourceLine" id="cb12-12" data-line-number="12"><span class="co">##  $ Species   : Factor w/ 3 levels &quot;Hsap&quot;,&quot;PanTro&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...</span></a>
<a class="sourceLine" id="cb12-13" data-line-number="13"><span class="co">##  $ Sex       : Factor w/ 2 levels &quot;F&quot;,&quot;M&quot;: 2 2 1 1 2 2 1 2 2 2 ...</span></a>
<a class="sourceLine" id="cb12-14" data-line-number="14"><span class="co">##  $ Age       : int  42 39 25 21 45 25 26 39 31 39 ...</span></a>
<a class="sourceLine" id="cb12-15" data-line-number="15"><span class="co">##  $ RIN       : num  9.8 6.4 8.8 7.9 6.6 5.8 8.7 6.6 8.1 2.7 ...</span></a>
<a class="sourceLine" id="cb12-16" data-line-number="16"><span class="co">##  $ Hemisphere: Factor w/ 2 levels &quot;L&quot;,&quot;R&quot;: 1 1 2 1 1 2 2 2 2 2 ...</span></a>
<a class="sourceLine" id="cb12-17" data-line-number="17"><span class="co">##  $ PMI       : int  13 11 18 6 16 5 19 18 10 18 ...</span></a></code></pre></div>
</div>
<div id="data-normalization" class="section level3">
<h3><span class="header-section-number">2.0.3</span> Data normalization</h3>
<p>Do you see something weird in the boxplot?</p>
<p>Pretty sure you realized that the distribution is skewed toward the bottom.</p>
<p>There are several factors that can influence this:</p>
<ul>
<li><p>Library size: sequencing depth.</p></li>
<li><p>Gene lenght: long gens = more reads</p></li>
<li><p>Library composition: some biological factors can influence the transcriptome between samples.</p></li>
<li><p>GC content: GC can influence mappability, therefore gene count.</p></li>
</ul>
<p>Several methods have been developped to normalize the values prior downstream analysis.
Here some examples:</p>
<ul>
<li><p><strong>CPM</strong> (<em>Counts per million</em>): no length considered</p></li>
<li><p><strong>RPKM</strong> (<em>Reads per kilobase per million</em>): length considered</p></li>
<li><p><strong>TPM</strong> (<em>Transcript per million</em>): length considered</p></li>
<li><p><strong>CQN</strong> (<em>Conditional quantile normalization</em>): length and GC content considered</p></li>
</ul>
<div id="calculate-cpm" class="section level4">
<h4><span class="header-section-number">2.0.3.1</span> Calculate CPM</h4>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="co"># Let&#39;s calculate the CPM</span></a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="kw">plan</span>(multiprocess)</a>
<a class="sourceLine" id="cb13-3" data-line-number="3"></a>
<a class="sourceLine" id="cb13-4" data-line-number="4"><span class="co"># future.apply is an alternative of the standard apply that</span></a>
<a class="sourceLine" id="cb13-5" data-line-number="5"><span class="co"># goes in parallel.  here we apply a function to the</span></a>
<a class="sourceLine" id="cb13-6" data-line-number="6"><span class="co"># expression matrix dividing each column by the total reads *</span></a>
<a class="sourceLine" id="cb13-7" data-line-number="7"><span class="co"># one million.</span></a>
<a class="sourceLine" id="cb13-8" data-line-number="8">cpm &lt;-<span class="st"> </span><span class="kw">future_apply</span>(exp, <span class="dv">2</span>, <span class="cf">function</span>(x) x<span class="op">/</span><span class="kw">sum</span>(<span class="kw">as.numeric</span>(x)) <span class="op">*</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-9" data-line-number="9"><span class="st">  </span><span class="dv">10</span><span class="op">^</span><span class="dv">6</span>)</a>
<a class="sourceLine" id="cb13-10" data-line-number="10"></a>
<a class="sourceLine" id="cb13-11" data-line-number="11"><span class="co"># Let&#39;s have a look!</span></a>
<a class="sourceLine" id="cb13-12" data-line-number="12"><span class="kw">head</span>(cpm)</a>
<a class="sourceLine" id="cb13-13" data-line-number="13"><span class="co">##          Hsap_1  Hsap_2 Hsap_3  Hsap_4 Hsap_5  Hsap_6</span></a>
<a class="sourceLine" id="cb13-14" data-line-number="14"><span class="co">## A1BG    3.65692 12.6344 7.4863 31.5862 9.3343 14.0957</span></a>
<a class="sourceLine" id="cb13-15" data-line-number="15"><span class="co">## A1CF    0.15421  0.3734 0.1582  0.0000 0.5834  0.1236</span></a>
<a class="sourceLine" id="cb13-16" data-line-number="16"><span class="co">## A2M     1.49801  4.1077 0.4218  0.0000 2.5457  1.5456</span></a>
<a class="sourceLine" id="cb13-17" data-line-number="17"><span class="co">## A2ML1   0.41856  0.9958 0.4218  0.2311 1.0607  0.6182</span></a>
<a class="sourceLine" id="cb13-18" data-line-number="18"><span class="co">## A3GALT2 1.93861  2.3651 1.9507  0.6934 0.5304  1.6692</span></a>
<a class="sourceLine" id="cb13-19" data-line-number="19"><span class="co">## A4GALT  0.02203  0.2490 0.1054  0.1541 0.0000  0.1236</span></a>
<a class="sourceLine" id="cb13-20" data-line-number="20"><span class="co">##         Hsap_7    Hsap_8  Hsap_9 Hsap_10 PanTro_1 PanTro_2</span></a>
<a class="sourceLine" id="cb13-21" data-line-number="21"><span class="co">## A1BG    5.6898 194.65722 14.0305  2.4087  11.1888  0.23248</span></a>
<a class="sourceLine" id="cb13-22" data-line-number="22"><span class="co">## A1CF    0.1962   0.06969  0.0000  0.0000   0.0000  0.00000</span></a>
<a class="sourceLine" id="cb13-23" data-line-number="23"><span class="co">## A2M     1.1118   0.62725  0.6829  0.4588   1.2432  4.92086</span></a>
<a class="sourceLine" id="cb13-24" data-line-number="24"><span class="co">## A2ML1   0.9810   3.13626  0.0000  0.4015   5.5944  0.11624</span></a>
<a class="sourceLine" id="cb13-25" data-line-number="25"><span class="co">## A3GALT2 3.2046  11.70870  1.3037  0.4588   3.9368  0.19373</span></a>
<a class="sourceLine" id="cb13-26" data-line-number="26"><span class="co">## A4GALT  0.1962   0.62725  0.2483  0.0000   0.2072  0.03875</span></a>
<a class="sourceLine" id="cb13-27" data-line-number="27"><span class="co">##         PanTro_3 PanTro_4 PanTro_5 PanTro_6 PanTro_7</span></a>
<a class="sourceLine" id="cb13-28" data-line-number="28"><span class="co">## A1BG      23.145  1.33764  0.74119   3.4020   7.3308</span></a>
<a class="sourceLine" id="cb13-29" data-line-number="29"><span class="co">## A1CF       0.000  0.06688  0.06177   0.0000   0.2115</span></a>
<a class="sourceLine" id="cb13-30" data-line-number="30"><span class="co">## A2M        5.545  0.13376  1.05002   1.8318   0.6344</span></a>
<a class="sourceLine" id="cb13-31" data-line-number="31"><span class="co">## A2ML1      0.000  0.53506  0.24706   0.3925   1.4098</span></a>
<a class="sourceLine" id="cb13-32" data-line-number="32"><span class="co">## A3GALT2    6.751  1.20388  0.43236   1.0468   4.1588</span></a>
<a class="sourceLine" id="cb13-33" data-line-number="33"><span class="co">## A4GALT     0.000  0.00000  0.00000   0.2617   0.4229</span></a>
<a class="sourceLine" id="cb13-34" data-line-number="34"><span class="co">##         PanTro_8 PanTro_9 PanTro_10 RheMac_1 RheMac_2</span></a>
<a class="sourceLine" id="cb13-35" data-line-number="35"><span class="co">## A1BG      0.7277   5.2930    4.3415   0.5717   1.1613</span></a>
<a class="sourceLine" id="cb13-36" data-line-number="36"><span class="co">## A1CF      0.3639   0.8318    0.0804   0.2144   0.1056</span></a>
<a class="sourceLine" id="cb13-37" data-line-number="37"><span class="co">## A2M       1.7153   1.1342    0.4020   0.6431   0.7390</span></a>
<a class="sourceLine" id="cb13-38" data-line-number="38"><span class="co">## A2ML1     0.6238   1.5879    1.2864   0.2858   0.5279</span></a>
<a class="sourceLine" id="cb13-39" data-line-number="39"><span class="co">## A3GALT2   0.6238   2.3440    4.8239   1.0004   0.9502</span></a>
<a class="sourceLine" id="cb13-40" data-line-number="40"><span class="co">## A4GALT    0.1559   0.3781    0.0000   0.2858   0.2112</span></a>
<a class="sourceLine" id="cb13-41" data-line-number="41"><span class="co">##         RheMac_3 RheMac_4 RheMac_5 RheMac_6 RheMac_7</span></a>
<a class="sourceLine" id="cb13-42" data-line-number="42"><span class="co">## A1BG      0.1875   0.1398   1.5296   0.0000   3.0919</span></a>
<a class="sourceLine" id="cb13-43" data-line-number="43"><span class="co">## A1CF      0.1875   0.0000   0.0000   0.0000   0.1066</span></a>
<a class="sourceLine" id="cb13-44" data-line-number="44"><span class="co">## A2M       2.2501   0.0000   1.9885   2.3751   0.7463</span></a>
<a class="sourceLine" id="cb13-45" data-line-number="45"><span class="co">## A2ML1     0.1875   0.6988   0.7648   0.1638   0.6397</span></a>
<a class="sourceLine" id="cb13-46" data-line-number="46"><span class="co">## A3GALT2   0.1875   0.0000   1.5296   0.4914   1.0662</span></a>
<a class="sourceLine" id="cb13-47" data-line-number="47"><span class="co">## A4GALT    0.0000   0.1398   0.0000   0.0819   0.3198</span></a>
<a class="sourceLine" id="cb13-48" data-line-number="48"><span class="co">##         RheMac_8 RheMac_9 RheMac_10</span></a>
<a class="sourceLine" id="cb13-49" data-line-number="49"><span class="co">## A1BG     0.06617   0.7008    0.6475</span></a>
<a class="sourceLine" id="cb13-50" data-line-number="50"><span class="co">## A1CF     0.06617   0.1752    0.0000</span></a>
<a class="sourceLine" id="cb13-51" data-line-number="51"><span class="co">## A2M      7.27894   2.2776    1.0792</span></a>
<a class="sourceLine" id="cb13-52" data-line-number="52"><span class="co">## A2ML1    0.52938   0.3504    0.0000</span></a>
<a class="sourceLine" id="cb13-53" data-line-number="53"><span class="co">## A3GALT2  0.99258   3.6793    0.7555</span></a>
<a class="sourceLine" id="cb13-54" data-line-number="54"><span class="co">## A4GALT   0.13234   0.0000    0.0000</span></a>
<a class="sourceLine" id="cb13-55" data-line-number="55"></a>
<a class="sourceLine" id="cb13-56" data-line-number="56"><span class="co"># The sum of each column is = 10^6.</span></a>
<a class="sourceLine" id="cb13-57" data-line-number="57"><span class="kw">colSums</span>(cpm)</a>
<a class="sourceLine" id="cb13-58" data-line-number="58"><span class="co">##    Hsap_1    Hsap_2    Hsap_3    Hsap_4    Hsap_5    Hsap_6 </span></a>
<a class="sourceLine" id="cb13-59" data-line-number="59"><span class="co">##     1e+06     1e+06     1e+06     1e+06     1e+06     1e+06 </span></a>
<a class="sourceLine" id="cb13-60" data-line-number="60"><span class="co">##    Hsap_7    Hsap_8    Hsap_9   Hsap_10  PanTro_1  PanTro_2 </span></a>
<a class="sourceLine" id="cb13-61" data-line-number="61"><span class="co">##     1e+06     1e+06     1e+06     1e+06     1e+06     1e+06 </span></a>
<a class="sourceLine" id="cb13-62" data-line-number="62"><span class="co">##  PanTro_3  PanTro_4  PanTro_5  PanTro_6  PanTro_7  PanTro_8 </span></a>
<a class="sourceLine" id="cb13-63" data-line-number="63"><span class="co">##     1e+06     1e+06     1e+06     1e+06     1e+06     1e+06 </span></a>
<a class="sourceLine" id="cb13-64" data-line-number="64"><span class="co">##  PanTro_9 PanTro_10  RheMac_1  RheMac_2  RheMac_3  RheMac_4 </span></a>
<a class="sourceLine" id="cb13-65" data-line-number="65"><span class="co">##     1e+06     1e+06     1e+06     1e+06     1e+06     1e+06 </span></a>
<a class="sourceLine" id="cb13-66" data-line-number="66"><span class="co">##  RheMac_5  RheMac_6  RheMac_7  RheMac_8  RheMac_9 RheMac_10 </span></a>
<a class="sourceLine" id="cb13-67" data-line-number="67"><span class="co">##     1e+06     1e+06     1e+06     1e+06     1e+06     1e+06</span></a></code></pre></div>
</div>
<div id="calculate-rpkm" class="section level4">
<h4><span class="header-section-number">2.0.3.2</span> Calculate RPKM</h4>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="co"># For RPKM you need the length of the genes (width)</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="kw">head</span>(width)</a>
<a class="sourceLine" id="cb14-3" data-line-number="3"><span class="co">##      Gene Length GCperc</span></a>
<a class="sourceLine" id="cb14-4" data-line-number="4"><span class="co">## 1    A1BG   4006 0.5580</span></a>
<a class="sourceLine" id="cb14-5" data-line-number="5"><span class="co">## 2    A1CF   9603 0.3624</span></a>
<a class="sourceLine" id="cb14-6" data-line-number="6"><span class="co">## 3     A2M   6384 0.3718</span></a>
<a class="sourceLine" id="cb14-7" data-line-number="7"><span class="co">## 4   A2ML1   7303 0.4423</span></a>
<a class="sourceLine" id="cb14-8" data-line-number="8"><span class="co">## 5 A3GALT2   1023 0.5419</span></a>
<a class="sourceLine" id="cb14-9" data-line-number="9"><span class="co">## 6  A4GALT   2943 0.5239</span></a>
<a class="sourceLine" id="cb14-10" data-line-number="10"></a>
<a class="sourceLine" id="cb14-11" data-line-number="11"><span class="co"># let&#39;s create a vector with the gene length</span></a>
<a class="sourceLine" id="cb14-12" data-line-number="12">l &lt;-<span class="st"> </span><span class="kw">as.vector</span>(width<span class="op">$</span>Length)</a>
<a class="sourceLine" id="cb14-13" data-line-number="13"></a>
<a class="sourceLine" id="cb14-14" data-line-number="14"><span class="co"># Calculate the RPKM</span></a>
<a class="sourceLine" id="cb14-15" data-line-number="15">rpkm &lt;-<span class="st"> </span><span class="kw">future_apply</span>(exp, <span class="dv">2</span>, <span class="cf">function</span>(x) <span class="dv">10</span><span class="op">^</span><span class="dv">9</span> <span class="op">*</span><span class="st"> </span>x<span class="op">/</span>l<span class="op">/</span><span class="kw">sum</span>(<span class="kw">as.numeric</span>(x)))</a>
<a class="sourceLine" id="cb14-16" data-line-number="16"></a>
<a class="sourceLine" id="cb14-17" data-line-number="17"><span class="co"># Let&#39;s have a look!</span></a>
<a class="sourceLine" id="cb14-18" data-line-number="18"><span class="kw">head</span>(rpkm)</a>
<a class="sourceLine" id="cb14-19" data-line-number="19"><span class="co">##           Hsap_1  Hsap_2  Hsap_3  Hsap_4  Hsap_5  Hsap_6</span></a>
<a class="sourceLine" id="cb14-20" data-line-number="20"><span class="co">## A1BG    0.912860 3.15386 1.86878 7.88473 2.33008 3.51865</span></a>
<a class="sourceLine" id="cb14-21" data-line-number="21"><span class="co">## A1CF    0.016058 0.03889 0.01647 0.00000 0.06075 0.01288</span></a>
<a class="sourceLine" id="cb14-22" data-line-number="22"><span class="co">## A2M     0.234651 0.64344 0.06607 0.00000 0.39877 0.24210</span></a>
<a class="sourceLine" id="cb14-23" data-line-number="23"><span class="co">## A2ML1   0.057314 0.13636 0.05775 0.03165 0.14524 0.08465</span></a>
<a class="sourceLine" id="cb14-24" data-line-number="24"><span class="co">## A3GALT2 1.895021 2.31188 1.90681 0.67777 0.51843 1.63170</span></a>
<a class="sourceLine" id="cb14-25" data-line-number="25"><span class="co">## A4GALT  0.007485 0.08459 0.03583 0.05235 0.00000 0.04201</span></a>
<a class="sourceLine" id="cb14-26" data-line-number="26"><span class="co">##          Hsap_7    Hsap_8  Hsap_9 Hsap_10 PanTro_1 PanTro_2</span></a>
<a class="sourceLine" id="cb14-27" data-line-number="27"><span class="co">## A1BG    1.42033 48.591418 3.50236 0.60128   2.7930  0.05803</span></a>
<a class="sourceLine" id="cb14-28" data-line-number="28"><span class="co">## A1CF    0.02043  0.007258 0.00000 0.00000   0.0000  0.00000</span></a>
<a class="sourceLine" id="cb14-29" data-line-number="29"><span class="co">## A2M     0.17416  0.098254 0.10697 0.07187   0.1947  0.77081</span></a>
<a class="sourceLine" id="cb14-30" data-line-number="30"><span class="co">## A2ML1   0.13433  0.429448 0.00000 0.05497   0.7660  0.01592</span></a>
<a class="sourceLine" id="cb14-31" data-line-number="31"><span class="co">## A3GALT2 3.13257 11.445459 1.27441 0.44849   3.8483  0.18938</span></a>
<a class="sourceLine" id="cb14-32" data-line-number="32"><span class="co">## A4GALT  0.06667  0.213134 0.08438 0.00000   0.0704  0.01317</span></a>
<a class="sourceLine" id="cb14-33" data-line-number="33"><span class="co">##         PanTro_3 PanTro_4 PanTro_5 PanTro_6 PanTro_7</span></a>
<a class="sourceLine" id="cb14-34" data-line-number="34"><span class="co">## A1BG      5.7777 0.333909 0.185019  0.84922  1.82995</span></a>
<a class="sourceLine" id="cb14-35" data-line-number="35"><span class="co">## A1CF      0.0000 0.006965 0.006432  0.00000  0.02202</span></a>
<a class="sourceLine" id="cb14-36" data-line-number="36"><span class="co">## A2M       0.8686 0.020953 0.164476  0.28694  0.09937</span></a>
<a class="sourceLine" id="cb14-37" data-line-number="37"><span class="co">## A2ML1     0.0000 0.073265 0.033830  0.05375  0.19304</span></a>
<a class="sourceLine" id="cb14-38" data-line-number="38"><span class="co">## A3GALT2   6.5989 1.176810 0.422639  1.02322  4.06531</span></a>
<a class="sourceLine" id="cb14-39" data-line-number="39"><span class="co">## A4GALT    0.0000 0.000000 0.000000  0.08892  0.14371</span></a>
<a class="sourceLine" id="cb14-40" data-line-number="40"><span class="co">##         PanTro_8 PanTro_9 PanTro_10 RheMac_1 RheMac_2</span></a>
<a class="sourceLine" id="cb14-41" data-line-number="41"><span class="co">## A1BG     0.18166  1.32126  1.083746  0.14270  0.28990</span></a>
<a class="sourceLine" id="cb14-42" data-line-number="42"><span class="co">## A1CF     0.03789  0.08661  0.008372  0.02232  0.01099</span></a>
<a class="sourceLine" id="cb14-43" data-line-number="43"><span class="co">## A2M      0.26869  0.17766  0.062968  0.10074  0.11576</span></a>
<a class="sourceLine" id="cb14-44" data-line-number="44"><span class="co">## A2ML1    0.08541  0.21743  0.176142  0.03914  0.07228</span></a>
<a class="sourceLine" id="cb14-45" data-line-number="45"><span class="co">## A3GALT2  0.60974  2.29133  4.715420  0.97791  0.92883</span></a>
<a class="sourceLine" id="cb14-46" data-line-number="46"><span class="co">## A4GALT   0.05299  0.12846  0.000000  0.09712  0.07175</span></a>
<a class="sourceLine" id="cb14-47" data-line-number="47"><span class="co">##         RheMac_3 RheMac_4 RheMac_5 RheMac_6 RheMac_7</span></a>
<a class="sourceLine" id="cb14-48" data-line-number="48"><span class="co">## A1BG     0.04681  0.03489   0.3818  0.00000  0.77181</span></a>
<a class="sourceLine" id="cb14-49" data-line-number="49"><span class="co">## A1CF     0.01953  0.00000   0.0000  0.00000  0.01110</span></a>
<a class="sourceLine" id="cb14-50" data-line-number="50"><span class="co">## A2M      0.35246  0.00000   0.3115  0.37203  0.11690</span></a>
<a class="sourceLine" id="cb14-51" data-line-number="51"><span class="co">## A2ML1    0.02568  0.09569   0.1047  0.02243  0.08759</span></a>
<a class="sourceLine" id="cb14-52" data-line-number="52"><span class="co">## A3GALT2  0.18329  0.00000   1.4952  0.48034  1.04219</span></a>
<a class="sourceLine" id="cb14-53" data-line-number="53"><span class="co">## A4GALT   0.00000  0.04749   0.0000  0.02783  0.10868</span></a>
<a class="sourceLine" id="cb14-54" data-line-number="54"><span class="co">##         RheMac_8 RheMac_9 RheMac_10</span></a>
<a class="sourceLine" id="cb14-55" data-line-number="55"><span class="co">## A1BG    0.016518  0.17494    0.1616</span></a>
<a class="sourceLine" id="cb14-56" data-line-number="56"><span class="co">## A1CF    0.006891  0.01824    0.0000</span></a>
<a class="sourceLine" id="cb14-57" data-line-number="57"><span class="co">## A2M     1.140185  0.35677    0.1691</span></a>
<a class="sourceLine" id="cb14-58" data-line-number="58"><span class="co">## A2ML1   0.072488  0.04798    0.0000</span></a>
<a class="sourceLine" id="cb14-59" data-line-number="59"><span class="co">## A3GALT2 0.970267  3.59655    0.7385</span></a>
<a class="sourceLine" id="cb14-60" data-line-number="60"><span class="co">## A4GALT  0.044969  0.00000    0.0000</span></a>
<a class="sourceLine" id="cb14-61" data-line-number="61"></a>
<a class="sourceLine" id="cb14-62" data-line-number="62"><span class="co"># Now the sum are all different!.</span></a>
<a class="sourceLine" id="cb14-63" data-line-number="63"><span class="kw">colSums</span>(rpkm)</a>
<a class="sourceLine" id="cb14-64" data-line-number="64"><span class="co">##    Hsap_1    Hsap_2    Hsap_3    Hsap_4    Hsap_5    Hsap_6 </span></a>
<a class="sourceLine" id="cb14-65" data-line-number="65"><span class="co">##    135992    137546    136729    136757    135507    137081 </span></a>
<a class="sourceLine" id="cb14-66" data-line-number="66"><span class="co">##    Hsap_7    Hsap_8    Hsap_9   Hsap_10  PanTro_1  PanTro_2 </span></a>
<a class="sourceLine" id="cb14-67" data-line-number="67"><span class="co">##    140837    140161    135588    134455    149958    133939 </span></a>
<a class="sourceLine" id="cb14-68" data-line-number="68"><span class="co">##  PanTro_3  PanTro_4  PanTro_5  PanTro_6  PanTro_7  PanTro_8 </span></a>
<a class="sourceLine" id="cb14-69" data-line-number="69"><span class="co">##    145104    135347    132978    138628    141954    135596 </span></a>
<a class="sourceLine" id="cb14-70" data-line-number="70"><span class="co">##  PanTro_9 PanTro_10  RheMac_1  RheMac_2  RheMac_3  RheMac_4 </span></a>
<a class="sourceLine" id="cb14-71" data-line-number="71"><span class="co">##    138648    139438    134781    142715    132472    139544 </span></a>
<a class="sourceLine" id="cb14-72" data-line-number="72"><span class="co">##  RheMac_5  RheMac_6  RheMac_7  RheMac_8  RheMac_9 RheMac_10 </span></a>
<a class="sourceLine" id="cb14-73" data-line-number="73"><span class="co">##    141869    133605    140033    136176    142478    132871</span></a></code></pre></div>
</div>
<div id="calculate-tpm" class="section level4">
<h4><span class="header-section-number">2.0.3.3</span> Calculate TPM</h4>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="co"># For TPM you need the length of the genes (l)</span></a>
<a class="sourceLine" id="cb15-2" data-line-number="2"></a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="co"># First step: Calculate the reads per kilobase</span></a>
<a class="sourceLine" id="cb15-4" data-line-number="4">rpk &lt;-<span class="st"> </span><span class="kw">future_apply</span>(exp, <span class="dv">2</span>, <span class="cf">function</span>(x) x<span class="op">/</span>(l<span class="op">/</span><span class="dv">1000</span>))</a>
<a class="sourceLine" id="cb15-5" data-line-number="5"></a>
<a class="sourceLine" id="cb15-6" data-line-number="6"><span class="co"># Now the TPM</span></a>
<a class="sourceLine" id="cb15-7" data-line-number="7">tpm &lt;-<span class="st"> </span><span class="kw">future_apply</span>(rpk, <span class="dv">2</span>, <span class="cf">function</span>(x) x<span class="op">/</span><span class="kw">sum</span>(<span class="kw">as.numeric</span>(x)) <span class="op">*</span><span class="st"> </span></a>
<a class="sourceLine" id="cb15-8" data-line-number="8"><span class="st">  </span><span class="dv">10</span><span class="op">^</span><span class="dv">6</span>)</a>
<a class="sourceLine" id="cb15-9" data-line-number="9"></a>
<a class="sourceLine" id="cb15-10" data-line-number="10"><span class="co"># Let&#39;s have a look!</span></a>
<a class="sourceLine" id="cb15-11" data-line-number="11"><span class="kw">head</span>(tpm)</a>
<a class="sourceLine" id="cb15-12" data-line-number="12"><span class="co">##           Hsap_1  Hsap_2  Hsap_3  Hsap_4  Hsap_5   Hsap_6</span></a>
<a class="sourceLine" id="cb15-13" data-line-number="13"><span class="co">## A1BG     6.71259 22.9295 13.6678 57.6550 17.1953 25.66842</span></a>
<a class="sourceLine" id="cb15-14" data-line-number="14"><span class="co">## A1CF     0.11808  0.2827  0.1205  0.0000  0.4483  0.09393</span></a>
<a class="sourceLine" id="cb15-15" data-line-number="15"><span class="co">## A2M      1.72548  4.6780  0.4832  0.0000  2.9428  1.76613</span></a>
<a class="sourceLine" id="cb15-16" data-line-number="16"><span class="co">## A2ML1    0.42145  0.9914  0.4224  0.2314  1.0719  0.61755</span></a>
<a class="sourceLine" id="cb15-17" data-line-number="17"><span class="co">## A3GALT2 13.93477 16.8080 13.9459  4.9560  3.8259 11.90319</span></a>
<a class="sourceLine" id="cb15-18" data-line-number="18"><span class="co">## A4GALT   0.05504  0.6150  0.2620  0.3828  0.0000  0.30649</span></a>
<a class="sourceLine" id="cb15-19" data-line-number="19"><span class="co">##          Hsap_7    Hsap_8  Hsap_9 Hsap_10 PanTro_1 PanTro_2</span></a>
<a class="sourceLine" id="cb15-20" data-line-number="20"><span class="co">## A1BG    10.0849 346.68275 25.8309  4.4720  18.6252   0.4333</span></a>
<a class="sourceLine" id="cb15-21" data-line-number="21"><span class="co">## A1CF     0.1451   0.05178  0.0000  0.0000   0.0000   0.0000</span></a>
<a class="sourceLine" id="cb15-22" data-line-number="22"><span class="co">## A2M      1.2366   0.70101  0.7889  0.5345   1.2986   5.7550</span></a>
<a class="sourceLine" id="cb15-23" data-line-number="23"><span class="co">## A2ML1    0.9538   3.06396  0.0000  0.4088   5.1083   0.1188</span></a>
<a class="sourceLine" id="cb15-24" data-line-number="24"><span class="co">## A3GALT2 22.2425  81.65935  9.3991  3.3356  25.6623   1.4139</span></a>
<a class="sourceLine" id="cb15-25" data-line-number="25"><span class="co">## A4GALT   0.4734   1.52063  0.6223  0.0000   0.4695   0.0983</span></a>
<a class="sourceLine" id="cb15-26" data-line-number="26"><span class="co">##         PanTro_3 PanTro_4 PanTro_5 PanTro_6 PanTro_7</span></a>
<a class="sourceLine" id="cb15-27" data-line-number="27"><span class="co">## A1BG      39.817  2.46706  1.39135   6.1259  12.8912</span></a>
<a class="sourceLine" id="cb15-28" data-line-number="28"><span class="co">## A1CF       0.000  0.05146  0.04837   0.0000   0.1551</span></a>
<a class="sourceLine" id="cb15-29" data-line-number="29"><span class="co">## A2M        5.986  0.15481  1.23687   2.0699   0.7000</span></a>
<a class="sourceLine" id="cb15-30" data-line-number="30"><span class="co">## A2ML1      0.000  0.54131  0.25441   0.3877   1.3599</span></a>
<a class="sourceLine" id="cb15-31" data-line-number="31"><span class="co">## A3GALT2   45.477  8.69476  3.17826   7.3811  28.6383</span></a>
<a class="sourceLine" id="cb15-32" data-line-number="32"><span class="co">## A4GALT     0.000  0.00000  0.00000   0.6414   1.0124</span></a>
<a class="sourceLine" id="cb15-33" data-line-number="33"><span class="co">##         PanTro_8 PanTro_9 PanTro_10 RheMac_1 RheMac_2</span></a>
<a class="sourceLine" id="cb15-34" data-line-number="34"><span class="co">## A1BG      1.3397   9.5296   7.77226   1.0588  2.03133</span></a>
<a class="sourceLine" id="cb15-35" data-line-number="35"><span class="co">## A1CF      0.2794   0.6247   0.06004   0.1656  0.07704</span></a>
<a class="sourceLine" id="cb15-36" data-line-number="36"><span class="co">## A2M       1.9816   1.2814   0.45159   0.7474  0.81116</span></a>
<a class="sourceLine" id="cb15-37" data-line-number="37"><span class="co">## A2ML1     0.6299   1.5682   1.26323   0.2904  0.50649</span></a>
<a class="sourceLine" id="cb15-38" data-line-number="38"><span class="co">## A3GALT2   4.4967  16.5262  33.81741   7.2556  6.50828</span></a>
<a class="sourceLine" id="cb15-39" data-line-number="39"><span class="co">## A4GALT    0.3908   0.9265   0.00000   0.7206  0.50274</span></a>
<a class="sourceLine" id="cb15-40" data-line-number="40"><span class="co">##         RheMac_3 RheMac_4 RheMac_5 RheMac_6 RheMac_7</span></a>
<a class="sourceLine" id="cb15-41" data-line-number="41"><span class="co">## A1BG      0.3533   0.2500   2.6915   0.0000  5.51160</span></a>
<a class="sourceLine" id="cb15-42" data-line-number="42"><span class="co">## A1CF      0.1474   0.0000   0.0000   0.0000  0.07928</span></a>
<a class="sourceLine" id="cb15-43" data-line-number="43"><span class="co">## A2M       2.6606   0.0000   2.1956   2.7846  0.83483</span></a>
<a class="sourceLine" id="cb15-44" data-line-number="44"><span class="co">## A2ML1     0.1938   0.6857   0.7382   0.1679  0.62552</span></a>
<a class="sourceLine" id="cb15-45" data-line-number="45"><span class="co">## A3GALT2   1.3836   0.0000  10.5396   3.5953  7.44243</span></a>
<a class="sourceLine" id="cb15-46" data-line-number="46"><span class="co">## A4GALT    0.0000   0.3403   0.0000   0.2083  0.77611</span></a>
<a class="sourceLine" id="cb15-47" data-line-number="47"><span class="co">##         RheMac_8 RheMac_9 RheMac_10</span></a>
<a class="sourceLine" id="cb15-48" data-line-number="48"><span class="co">## A1BG      0.1213   1.2278     1.217</span></a>
<a class="sourceLine" id="cb15-49" data-line-number="49"><span class="co">## A1CF      0.0506   0.1281     0.000</span></a>
<a class="sourceLine" id="cb15-50" data-line-number="50"><span class="co">## A2M       8.3729   2.5041     1.272</span></a>
<a class="sourceLine" id="cb15-51" data-line-number="51"><span class="co">## A2ML1     0.5323   0.3368     0.000</span></a>
<a class="sourceLine" id="cb15-52" data-line-number="52"><span class="co">## A3GALT2   7.1251  25.2428     5.558</span></a>
<a class="sourceLine" id="cb15-53" data-line-number="53"><span class="co">## A4GALT    0.3302   0.0000     0.000</span></a>
<a class="sourceLine" id="cb15-54" data-line-number="54"></a>
<a class="sourceLine" id="cb15-55" data-line-number="55"><span class="co"># The sum of each column is = 10^6!.</span></a>
<a class="sourceLine" id="cb15-56" data-line-number="56"><span class="kw">colSums</span>(tpm)</a>
<a class="sourceLine" id="cb15-57" data-line-number="57"><span class="co">##    Hsap_1    Hsap_2    Hsap_3    Hsap_4    Hsap_5    Hsap_6 </span></a>
<a class="sourceLine" id="cb15-58" data-line-number="58"><span class="co">##     1e+06     1e+06     1e+06     1e+06     1e+06     1e+06 </span></a>
<a class="sourceLine" id="cb15-59" data-line-number="59"><span class="co">##    Hsap_7    Hsap_8    Hsap_9   Hsap_10  PanTro_1  PanTro_2 </span></a>
<a class="sourceLine" id="cb15-60" data-line-number="60"><span class="co">##     1e+06     1e+06     1e+06     1e+06     1e+06     1e+06 </span></a>
<a class="sourceLine" id="cb15-61" data-line-number="61"><span class="co">##  PanTro_3  PanTro_4  PanTro_5  PanTro_6  PanTro_7  PanTro_8 </span></a>
<a class="sourceLine" id="cb15-62" data-line-number="62"><span class="co">##     1e+06     1e+06     1e+06     1e+06     1e+06     1e+06 </span></a>
<a class="sourceLine" id="cb15-63" data-line-number="63"><span class="co">##  PanTro_9 PanTro_10  RheMac_1  RheMac_2  RheMac_3  RheMac_4 </span></a>
<a class="sourceLine" id="cb15-64" data-line-number="64"><span class="co">##     1e+06     1e+06     1e+06     1e+06     1e+06     1e+06 </span></a>
<a class="sourceLine" id="cb15-65" data-line-number="65"><span class="co">##  RheMac_5  RheMac_6  RheMac_7  RheMac_8  RheMac_9 RheMac_10 </span></a>
<a class="sourceLine" id="cb15-66" data-line-number="66"><span class="co">##     1e+06     1e+06     1e+06     1e+06     1e+06     1e+06</span></a></code></pre></div>
</div>
<div id="calculate-cqn" class="section level4">
<h4><span class="header-section-number">2.0.3.4</span> Calculate CQN</h4>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw">suppressPackageStartupMessages</span>(<span class="kw">library</span>(cqn))</a>
<a class="sourceLine" id="cb16-2" data-line-number="2"></a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="co"># let&#39;s create a vector with the gene length and gc content</span></a>
<a class="sourceLine" id="cb16-4" data-line-number="4">length &lt;-<span class="st"> </span><span class="kw">as.vector</span>(width<span class="op">$</span>Length)</a>
<a class="sourceLine" id="cb16-5" data-line-number="5">gc_content &lt;-<span class="st"> </span><span class="kw">as.vector</span>(width<span class="op">$</span>GCperc)</a>
<a class="sourceLine" id="cb16-6" data-line-number="6"></a>
<a class="sourceLine" id="cb16-7" data-line-number="7"><span class="co"># Calculate CQN and normalized values</span></a>
<a class="sourceLine" id="cb16-8" data-line-number="8">temp &lt;-<span class="st"> </span><span class="kw">cqn</span>(exp, <span class="dt">lengths =</span> length, <span class="dt">x =</span> gc_content, <span class="dt">sizeFactors =</span> <span class="kw">colSums</span>(exp), </a>
<a class="sourceLine" id="cb16-9" data-line-number="9">  <span class="dt">lengthMethod =</span> <span class="st">&quot;fixed&quot;</span>, <span class="dt">sqn =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb16-10" data-line-number="10"></a>
<a class="sourceLine" id="cb16-11" data-line-number="11"><span class="co"># get log2(normalized RPKM) values</span></a>
<a class="sourceLine" id="cb16-12" data-line-number="12">quantGC &lt;-<span class="st"> </span><span class="dv">2</span><span class="op">^</span>(temp<span class="op">$</span>y <span class="op">+</span><span class="st"> </span>temp<span class="op">$</span>offset)</a>
<a class="sourceLine" id="cb16-13" data-line-number="13"></a>
<a class="sourceLine" id="cb16-14" data-line-number="14"><span class="co"># Let&#39;s have a look!</span></a>
<a class="sourceLine" id="cb16-15" data-line-number="15"><span class="kw">head</span>(quantGC)</a>
<a class="sourceLine" id="cb16-16" data-line-number="16"><span class="co">##          Hsap_1   Hsap_2  Hsap_3    Hsap_4  Hsap_5  Hsap_6</span></a>
<a class="sourceLine" id="cb16-17" data-line-number="17"><span class="co">## A1BG    2.56927 10.55845 5.08924 30.841415 9.74631 8.55624</span></a>
<a class="sourceLine" id="cb16-18" data-line-number="18"><span class="co">## A1CF    0.01205  0.02649 0.01316  0.004706 0.03552 0.01240</span></a>
<a class="sourceLine" id="cb16-19" data-line-number="19"><span class="co">## A2M     0.15803  0.39396 0.04539  0.007449 0.23461 0.16546</span></a>
<a class="sourceLine" id="cb16-20" data-line-number="20"><span class="co">## A2ML1   0.06277  0.15799 0.06846  0.054177 0.18876 0.09853</span></a>
<a class="sourceLine" id="cb16-21" data-line-number="21"><span class="co">## A3GALT2 4.70302  6.63554 4.55666  2.526362 1.98350 3.65302</span></a>
<a class="sourceLine" id="cb16-22" data-line-number="22"><span class="co">## A4GALT  0.03149  0.24275 0.10567  0.220901 0.05107 0.11910</span></a>
<a class="sourceLine" id="cb16-23" data-line-number="23"><span class="co">##          Hsap_7   Hsap_8    Hsap_9  Hsap_10 PanTro_1</span></a>
<a class="sourceLine" id="cb16-24" data-line-number="24"><span class="co">## A1BG    1.52687 156.7761 15.730795 2.678769  1.68308</span></a>
<a class="sourceLine" id="cb16-25" data-line-number="25"><span class="co">## A1CF    0.01873   0.0109  0.004176 0.003558  0.03280</span></a>
<a class="sourceLine" id="cb16-26" data-line-number="26"><span class="co">## A2M     0.12593   0.0843  0.080090 0.050280  0.30675</span></a>
<a class="sourceLine" id="cb16-27" data-line-number="27"><span class="co">## A2ML1   0.12394   0.5023  0.011012 0.078499  0.64153</span></a>
<a class="sourceLine" id="cb16-28" data-line-number="28"><span class="co">## A3GALT2 3.36678  30.5082  4.944737 1.853574  2.52037</span></a>
<a class="sourceLine" id="cb16-29" data-line-number="29"><span class="co">## A4GALT  0.09204   0.5077  0.314778 0.058764  0.09251</span></a>
<a class="sourceLine" id="cb16-30" data-line-number="30"><span class="co">##         PanTro_2 PanTro_3 PanTro_4 PanTro_5 PanTro_6</span></a>
<a class="sourceLine" id="cb16-31" data-line-number="31"><span class="co">## A1BG    0.323282  7.57013 1.203895 1.307150  2.22728</span></a>
<a class="sourceLine" id="cb16-32" data-line-number="32"><span class="co">## A1CF    0.002187  0.02701 0.008085 0.006478  0.01398</span></a>
<a class="sourceLine" id="cb16-33" data-line-number="33"><span class="co">## A2M     0.456473  0.90128 0.018918 0.093457  0.29765</span></a>
<a class="sourceLine" id="cb16-34" data-line-number="34"><span class="co">## A2ML1   0.032910  0.02960 0.100878 0.059772  0.08046</span></a>
<a class="sourceLine" id="cb16-35" data-line-number="35"><span class="co">## A3GALT2 0.968989  8.16268 3.743524 2.591185  2.57671</span></a>
<a class="sourceLine" id="cb16-36" data-line-number="36"><span class="co">## A4GALT  0.097419  0.08978 0.058774 0.089540  0.25963</span></a>
<a class="sourceLine" id="cb16-37" data-line-number="37"><span class="co">##         PanTro_7 PanTro_8 PanTro_9 PanTro_10 RheMac_1</span></a>
<a class="sourceLine" id="cb16-38" data-line-number="38"><span class="co">## A1BG     1.36851  0.80972  1.83918   1.15951  0.47642</span></a>
<a class="sourceLine" id="cb16-39" data-line-number="39"><span class="co">## A1CF     0.02369  0.02137  0.08902   0.01406  0.01753</span></a>
<a class="sourceLine" id="cb16-40" data-line-number="40"><span class="co">## A2M      0.08660  0.14582  0.17202   0.06197  0.06846</span></a>
<a class="sourceLine" id="cb16-41" data-line-number="41"><span class="co">## A2ML1    0.14986  0.11238  0.22794   0.17717  0.05387</span></a>
<a class="sourceLine" id="cb16-42" data-line-number="42"><span class="co">## A3GALT2  3.05732  2.36518  3.10524   5.06858  2.79150</span></a>
<a class="sourceLine" id="cb16-43" data-line-number="43"><span class="co">## A4GALT   0.12378  0.21146  0.19201   0.02893  0.28296</span></a>
<a class="sourceLine" id="cb16-44" data-line-number="44"><span class="co">##         RheMac_2 RheMac_3 RheMac_4 RheMac_5 RheMac_6</span></a>
<a class="sourceLine" id="cb16-45" data-line-number="45"><span class="co">## A1BG     0.24532  0.38822  0.09098  0.35784 0.073764</span></a>
<a class="sourceLine" id="cb16-46" data-line-number="46"><span class="co">## A1CF     0.02100  0.01422  0.01176  0.01670 0.005465</span></a>
<a class="sourceLine" id="cb16-47" data-line-number="47"><span class="co">## A2M      0.11888  0.19411  0.01709  0.32997 0.255428</span></a>
<a class="sourceLine" id="cb16-48" data-line-number="48"><span class="co">## A2ML1    0.06967  0.05314  0.10073  0.10292 0.042981</span></a>
<a class="sourceLine" id="cb16-49" data-line-number="49"><span class="co">## A3GALT2  0.80543  1.29033  0.16979  1.39610 1.816089</span></a>
<a class="sourceLine" id="cb16-50" data-line-number="50"><span class="co">## A4GALT   0.08459  0.12283  0.11129  0.04386 0.157588</span></a>
<a class="sourceLine" id="cb16-51" data-line-number="51"><span class="co">##         RheMac_7 RheMac_8 RheMac_9 RheMac_10</span></a>
<a class="sourceLine" id="cb16-52" data-line-number="52"><span class="co">## A1BG     0.66299  0.05289  0.17867  1.095789</span></a>
<a class="sourceLine" id="cb16-53" data-line-number="53"><span class="co">## A1CF     0.02118  0.01095  0.04853  0.005946</span></a>
<a class="sourceLine" id="cb16-54" data-line-number="54"><span class="co">## A2M      0.12278  0.92096  0.46590  0.106475</span></a>
<a class="sourceLine" id="cb16-55" data-line-number="55"><span class="co">## A2ML1    0.08463  0.09071  0.06470  0.021158</span></a>
<a class="sourceLine" id="cb16-56" data-line-number="56"><span class="co">## A3GALT2  0.95892  1.61560  3.16347  4.244491</span></a>
<a class="sourceLine" id="cb16-57" data-line-number="57"><span class="co">## A4GALT   0.12191  0.10148  0.05126  0.153846</span></a>
<a class="sourceLine" id="cb16-58" data-line-number="58"></a>
<a class="sourceLine" id="cb16-59" data-line-number="59"><span class="co"># The sum of each column is = 10^6!.</span></a>
<a class="sourceLine" id="cb16-60" data-line-number="60"><span class="kw">colSums</span>(quantGC)</a>
<a class="sourceLine" id="cb16-61" data-line-number="61"><span class="co">##    Hsap_1    Hsap_2    Hsap_3    Hsap_4    Hsap_5    Hsap_6 </span></a>
<a class="sourceLine" id="cb16-62" data-line-number="62"><span class="co">##    136938    135675    132641    148347    140195    136839 </span></a>
<a class="sourceLine" id="cb16-63" data-line-number="63"><span class="co">##    Hsap_7    Hsap_8    Hsap_9   Hsap_10  PanTro_1  PanTro_2 </span></a>
<a class="sourceLine" id="cb16-64" data-line-number="64"><span class="co">##    117928    159501    155370    143235    133363    149459 </span></a>
<a class="sourceLine" id="cb16-65" data-line-number="65"><span class="co">##  PanTro_3  PanTro_4  PanTro_5  PanTro_6  PanTro_7  PanTro_8 </span></a>
<a class="sourceLine" id="cb16-66" data-line-number="66"><span class="co">##    147426    134711    138760    165924    108437    128301 </span></a>
<a class="sourceLine" id="cb16-67" data-line-number="67"><span class="co">##  PanTro_9 PanTro_10  RheMac_1  RheMac_2  RheMac_3  RheMac_4 </span></a>
<a class="sourceLine" id="cb16-68" data-line-number="68"><span class="co">##    143486    132448    132904    120266    134040    128488 </span></a>
<a class="sourceLine" id="cb16-69" data-line-number="69"><span class="co">##  RheMac_5  RheMac_6  RheMac_7  RheMac_8  RheMac_9 RheMac_10 </span></a>
<a class="sourceLine" id="cb16-70" data-line-number="70"><span class="co">##    127934    145847    122743    142587    140086    137997</span></a></code></pre></div>
</div>
<div id="check-the-normalized-data" class="section level4">
<h4><span class="header-section-number">2.0.3.5</span> Check the normalized data</h4>
<p>Now we can have a look to the distribution of the normalized data</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co"># Let&#39;s color based on the species</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2">colors &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">factor</span>(demo<span class="op">$</span>Species)) <span class="op">+</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb17-3" data-line-number="3"></a>
<a class="sourceLine" id="cb17-4" data-line-number="4"><span class="co"># Check the boxplot for CPM</span></a>
<a class="sourceLine" id="cb17-5" data-line-number="5"></a>
<a class="sourceLine" id="cb17-6" data-line-number="6"><span class="kw">boxplot</span>(cpm)</a></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/normDistrib-1.png" width="65%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"></a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="co"># Still skewed? Let&#39;s make every normally distributed the +1</span></a>
<a class="sourceLine" id="cb18-3" data-line-number="3"><span class="co"># is a offset reads to keep the 0 as 0!</span></a>
<a class="sourceLine" id="cb18-4" data-line-number="4"><span class="kw">boxplot</span>(<span class="kw">log2</span>(cpm <span class="op">+</span><span class="st"> </span><span class="dv">1</span>), <span class="dt">notch =</span> <span class="ot">TRUE</span>, <span class="dt">main =</span> <span class="st">&quot;log2(CPM+1)&quot;</span>, <span class="dt">col =</span> colors)</a>
<a class="sourceLine" id="cb18-5" data-line-number="5"><span class="kw">abline</span>(<span class="dt">h =</span> <span class="kw">median</span>(<span class="kw">log2</span>(cpm <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)), <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>)</a></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/normDistrib-2.png" width="65%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"></a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw">boxplot</span>(<span class="kw">log2</span>(rpkm <span class="op">+</span><span class="st"> </span><span class="dv">1</span>), <span class="dt">notch =</span> <span class="ot">TRUE</span>, <span class="dt">main =</span> <span class="st">&quot;log2(RPKM+1)&quot;</span>, </a>
<a class="sourceLine" id="cb19-3" data-line-number="3">  <span class="dt">col =</span> colors)</a>
<a class="sourceLine" id="cb19-4" data-line-number="4"><span class="kw">abline</span>(<span class="dt">h =</span> <span class="kw">median</span>(<span class="kw">log2</span>(rpkm <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)), <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>)</a></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/normDistrib-3.png" width="65%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"></a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="kw">boxplot</span>(<span class="kw">log2</span>(tpm <span class="op">+</span><span class="st"> </span><span class="dv">1</span>), <span class="dt">notch =</span> <span class="ot">TRUE</span>, <span class="dt">main =</span> <span class="st">&quot;log2(TPM+1)&quot;</span>, <span class="dt">col =</span> colors)</a>
<a class="sourceLine" id="cb20-3" data-line-number="3"><span class="kw">abline</span>(<span class="dt">h =</span> <span class="kw">median</span>(<span class="kw">log2</span>(tpm <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)), <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>)</a></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/normDistrib-4.png" width="65%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"></a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="kw">boxplot</span>(<span class="kw">log2</span>(quantGC <span class="op">+</span><span class="st"> </span><span class="dv">1</span>), <span class="dt">notch =</span> <span class="ot">TRUE</span>, <span class="dt">main =</span> <span class="st">&quot;log2(quantGC+1)&quot;</span>, </a>
<a class="sourceLine" id="cb21-3" data-line-number="3">  <span class="dt">col =</span> colors)</a>
<a class="sourceLine" id="cb21-4" data-line-number="4"><span class="kw">abline</span>(<span class="dt">h =</span> <span class="kw">median</span>(<span class="kw">log2</span>(quantGC <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)), <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>)</a></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/normDistrib-5.png" width="65%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="initial-data-exploration" class="section level3">
<h3><span class="header-section-number">2.0.4</span> Initial Data Exploration</h3>
<p>Now the counts are normalized and we know what we are dealing with, we can start to understand how the samples are similar/dissimilar based on the quantified gene expression profiles of the genes.
We would expect that biological/technical replicates will cluster together.
This can be computed with different unsupervised clustering methods (e.g. hierarchical clustering) or with dimensionality reductions methods (e.g. Principal Component Analysis)</p>
<p>Let’s start:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="co"># First let&#39;s make a matric with log2 scaled data.</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2">mat &lt;-<span class="st"> </span><span class="kw">log2</span>(cpm <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb22-3" data-line-number="3"></a>
<a class="sourceLine" id="cb22-4" data-line-number="4"><span class="co"># Calculate the variance for each gene This will detect the</span></a>
<a class="sourceLine" id="cb22-5" data-line-number="5"><span class="co"># most variables genes</span></a>
<a class="sourceLine" id="cb22-6" data-line-number="6">variance &lt;-<span class="st"> </span><span class="kw">apply</span>(mat, <span class="dv">1</span>, var)</a>
<a class="sourceLine" id="cb22-7" data-line-number="7"></a>
<a class="sourceLine" id="cb22-8" data-line-number="8"><span class="co"># Let&#39;s sort them and select the top 100 genes.</span></a>
<a class="sourceLine" id="cb22-9" data-line-number="9">selgene &lt;-<span class="st"> </span><span class="kw">names</span>(variance[<span class="kw">order</span>(variance, <span class="dt">decreasing =</span> T)][<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>])</a>
<a class="sourceLine" id="cb22-10" data-line-number="10"></a>
<a class="sourceLine" id="cb22-11" data-line-number="11"><span class="co"># Let&#39;s now calculate the Principal Components for the top</span></a>
<a class="sourceLine" id="cb22-12" data-line-number="12"><span class="co"># 100 genes</span></a>
<a class="sourceLine" id="cb22-13" data-line-number="13">pca_cpm &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(mat[selgene, ]))</a>
<a class="sourceLine" id="cb22-14" data-line-number="14"></a>
<a class="sourceLine" id="cb22-15" data-line-number="15"><span class="co"># Print the summary of the pcas</span></a>
<a class="sourceLine" id="cb22-16" data-line-number="16"><span class="kw">summary</span>(pca_cpm)</a>
<a class="sourceLine" id="cb22-17" data-line-number="17"><span class="co">## Importance of components:</span></a>
<a class="sourceLine" id="cb22-18" data-line-number="18"><span class="co">##                          PC1   PC2   PC3    PC4     PC5</span></a>
<a class="sourceLine" id="cb22-19" data-line-number="19"><span class="co">## Standard deviation     16.66 9.494 7.416 2.6041 1.85325</span></a>
<a class="sourceLine" id="cb22-20" data-line-number="20"><span class="co">## Proportion of Variance  0.61 0.198 0.121 0.0149 0.00754</span></a>
<a class="sourceLine" id="cb22-21" data-line-number="21"><span class="co">## Cumulative Proportion   0.61 0.808 0.929 0.9434 0.95096</span></a>
<a class="sourceLine" id="cb22-22" data-line-number="22"><span class="co">##                            PC6     PC7     PC8     PC9</span></a>
<a class="sourceLine" id="cb22-23" data-line-number="23"><span class="co">## Standard deviation     1.69094 1.57609 1.52093 1.48110</span></a>
<a class="sourceLine" id="cb22-24" data-line-number="24"><span class="co">## Proportion of Variance 0.00628 0.00546 0.00508 0.00482</span></a>
<a class="sourceLine" id="cb22-25" data-line-number="25"><span class="co">## Cumulative Proportion  0.95724 0.96270 0.96778 0.97259</span></a>
<a class="sourceLine" id="cb22-26" data-line-number="26"><span class="co">##                           PC10   PC11    PC12    PC13</span></a>
<a class="sourceLine" id="cb22-27" data-line-number="27"><span class="co">## Standard deviation     1.30788 1.2250 1.14155 1.07289</span></a>
<a class="sourceLine" id="cb22-28" data-line-number="28"><span class="co">## Proportion of Variance 0.00376 0.0033 0.00286 0.00253</span></a>
<a class="sourceLine" id="cb22-29" data-line-number="29"><span class="co">## Cumulative Proportion  0.97635 0.9796 0.98251 0.98503</span></a>
<a class="sourceLine" id="cb22-30" data-line-number="30"><span class="co">##                           PC14    PC15    PC16    PC17</span></a>
<a class="sourceLine" id="cb22-31" data-line-number="31"><span class="co">## Standard deviation     0.96525 0.91623 0.86372 0.80614</span></a>
<a class="sourceLine" id="cb22-32" data-line-number="32"><span class="co">## Proportion of Variance 0.00205 0.00184 0.00164 0.00143</span></a>
<a class="sourceLine" id="cb22-33" data-line-number="33"><span class="co">## Cumulative Proportion  0.98708 0.98892 0.99056 0.99199</span></a>
<a class="sourceLine" id="cb22-34" data-line-number="34"><span class="co">##                          PC18   PC19    PC20    PC21</span></a>
<a class="sourceLine" id="cb22-35" data-line-number="35"><span class="co">## Standard deviation     0.7399 0.7074 0.63831 0.63085</span></a>
<a class="sourceLine" id="cb22-36" data-line-number="36"><span class="co">## Proportion of Variance 0.0012 0.0011 0.00089 0.00087</span></a>
<a class="sourceLine" id="cb22-37" data-line-number="37"><span class="co">## Cumulative Proportion  0.9932 0.9943 0.99519 0.99606</span></a>
<a class="sourceLine" id="cb22-38" data-line-number="38"><span class="co">##                           PC22    PC23   PC24   PC25</span></a>
<a class="sourceLine" id="cb22-39" data-line-number="39"><span class="co">## Standard deviation     0.58808 0.54664 0.5226 0.4790</span></a>
<a class="sourceLine" id="cb22-40" data-line-number="40"><span class="co">## Proportion of Variance 0.00076 0.00066 0.0006 0.0005</span></a>
<a class="sourceLine" id="cb22-41" data-line-number="41"><span class="co">## Cumulative Proportion  0.99682 0.99747 0.9981 0.9986</span></a>
<a class="sourceLine" id="cb22-42" data-line-number="42"><span class="co">##                           PC26    PC27    PC28    PC29</span></a>
<a class="sourceLine" id="cb22-43" data-line-number="43"><span class="co">## Standard deviation     0.46119 0.43126 0.37833 0.32475</span></a>
<a class="sourceLine" id="cb22-44" data-line-number="44"><span class="co">## Proportion of Variance 0.00047 0.00041 0.00031 0.00023</span></a>
<a class="sourceLine" id="cb22-45" data-line-number="45"><span class="co">## Cumulative Proportion  0.99905 0.99945 0.99977 1.00000</span></a>
<a class="sourceLine" id="cb22-46" data-line-number="46"><span class="co">##                           PC30</span></a>
<a class="sourceLine" id="cb22-47" data-line-number="47"><span class="co">## Standard deviation     7.4e-15</span></a>
<a class="sourceLine" id="cb22-48" data-line-number="48"><span class="co">## Proportion of Variance 0.0e+00</span></a>
<a class="sourceLine" id="cb22-49" data-line-number="49"><span class="co">## Cumulative Proportion  1.0e+00</span></a>
<a class="sourceLine" id="cb22-50" data-line-number="50"></a>
<a class="sourceLine" id="cb22-51" data-line-number="51"><span class="co"># Let&#39;s have a look which PCA explain more variance</span></a>
<a class="sourceLine" id="cb22-52" data-line-number="52"><span class="kw">screeplot</span>(pca_cpm, <span class="dt">npcs =</span> <span class="dv">15</span>, <span class="dt">type =</span> <span class="st">&quot;barplot&quot;</span>)</a></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/pcas-1.png" width="65%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"></a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="co"># Now let&#39;s create a data frame with the PCAs and the values</span></a>
<a class="sourceLine" id="cb23-3" data-line-number="3"><span class="co"># for coloring</span></a>
<a class="sourceLine" id="cb23-4" data-line-number="4">PCi &lt;-<span class="st"> </span><span class="kw">data.frame</span>(pca_cpm<span class="op">$</span>x, <span class="dt">Species =</span> demo<span class="op">$</span>Species)</a>
<a class="sourceLine" id="cb23-5" data-line-number="5"></a>
<a class="sourceLine" id="cb23-6" data-line-number="6"><span class="co"># Let&#39;s calcualte the variance explained by the each</span></a>
<a class="sourceLine" id="cb23-7" data-line-number="7"><span class="co"># components</span></a>
<a class="sourceLine" id="cb23-8" data-line-number="8">eig &lt;-<span class="st"> </span>(pca_cpm<span class="op">$</span>sdev)<span class="op">^</span><span class="dv">2</span></a>
<a class="sourceLine" id="cb23-9" data-line-number="9">variance &lt;-<span class="st"> </span>eig <span class="op">*</span><span class="st"> </span><span class="dv">100</span><span class="op">/</span><span class="kw">sum</span>(eig)</a>
<a class="sourceLine" id="cb23-10" data-line-number="10"></a>
<a class="sourceLine" id="cb23-11" data-line-number="11"><span class="co"># Now plot the PCA</span></a>
<a class="sourceLine" id="cb23-12" data-line-number="12"><span class="kw">ggscatter</span>(PCi, <span class="dt">x =</span> <span class="st">&quot;PC1&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;PC2&quot;</span>, <span class="dt">color =</span> <span class="st">&quot;Species&quot;</span>, <span class="dt">palette =</span> <span class="kw">c</span>(<span class="st">&quot;red&quot;</span>, </a>
<a class="sourceLine" id="cb23-13" data-line-number="13">  <span class="st">&quot;grey60&quot;</span>, <span class="st">&quot;green&quot;</span>), <span class="dt">shape =</span> <span class="dv">21</span>, <span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">ellipse =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb23-14" data-line-number="14">  <span class="dt">mean.point =</span> <span class="ot">TRUE</span>, <span class="dt">star.plot =</span> <span class="ot">TRUE</span>) <span class="op">+</span><span class="st"> </span><span class="kw">xlab</span>(<span class="kw">paste</span>(<span class="st">&quot;PC1 (&quot;</span>, </a>
<a class="sourceLine" id="cb23-15" data-line-number="15">  <span class="kw">round</span>(variance[<span class="dv">1</span>], <span class="dv">1</span>), <span class="st">&quot;% )&quot;</span>)) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="kw">paste</span>(<span class="st">&quot;PC2 (&quot;</span>, <span class="kw">round</span>(variance[<span class="dv">2</span>], </a>
<a class="sourceLine" id="cb23-16" data-line-number="16">  <span class="dv">1</span>), <span class="st">&quot;% )&quot;</span>)) <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>()</a></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/pcas-2.png" width="65%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"></a>
<a class="sourceLine" id="cb24-2" data-line-number="2"><span class="co"># Now let&#39;s make a dendrogram based on hirachical clustering</span></a>
<a class="sourceLine" id="cb24-3" data-line-number="3">hc &lt;-<span class="st"> </span><span class="kw">hclust</span>(<span class="kw">dist</span>(<span class="kw">t</span>(mat[selgene, ])), <span class="st">&quot;ave&quot;</span>)</a>
<a class="sourceLine" id="cb24-4" data-line-number="4"><span class="kw">plot</span>(hc, <span class="dt">hang =</span> <span class="dv">-1</span>)</a></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/pcas-3.png" width="65%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"></a>
<a class="sourceLine" id="cb25-2" data-line-number="2"><span class="co"># Another way with correlation</span></a>
<a class="sourceLine" id="cb25-3" data-line-number="3">correlation_mat &lt;-<span class="st"> </span><span class="kw">cor</span>(mat[selgene, ], <span class="dt">method =</span> <span class="st">&quot;pearson&quot;</span>)</a>
<a class="sourceLine" id="cb25-4" data-line-number="4"><span class="kw">pheatmap</span>(correlation_mat, <span class="dt">annotation_col =</span> demo, <span class="dt">scale =</span> <span class="st">&quot;row&quot;</span>)</a></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/pcas-4.png" width="65%" style="display: block; margin: auto;" /></p>
</div>
<div id="variance-explained-by-covariates" class="section level3">
<h3><span class="header-section-number">2.0.5</span> Variance explained by covariates</h3>
<p>Now the big question: how much of the gene expression is explained by the biological/technical covaraites?
The covaraites can introduces systematic shifts in the downstream analysis and a good quality check is indeed to evaluate the variance that is explained by each of these covariates.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="co"># Re-check the demographic data</span></a>
<a class="sourceLine" id="cb26-2" data-line-number="2"><span class="kw">head</span>(demo)</a>
<a class="sourceLine" id="cb26-3" data-line-number="3"><span class="co">##        Species Sex Age RIN Hemisphere PMI</span></a>
<a class="sourceLine" id="cb26-4" data-line-number="4"><span class="co">## Hsap_1    Hsap   M  42 9.8          L  13</span></a>
<a class="sourceLine" id="cb26-5" data-line-number="5"><span class="co">## Hsap_2    Hsap   M  39 6.4          L  11</span></a>
<a class="sourceLine" id="cb26-6" data-line-number="6"><span class="co">## Hsap_3    Hsap   F  25 8.8          R  18</span></a>
<a class="sourceLine" id="cb26-7" data-line-number="7"><span class="co">## Hsap_4    Hsap   F  21 7.9          L   6</span></a>
<a class="sourceLine" id="cb26-8" data-line-number="8"><span class="co">## Hsap_5    Hsap   M  45 6.6          L  16</span></a>
<a class="sourceLine" id="cb26-9" data-line-number="9"><span class="co">## Hsap_6    Hsap   M  25 5.8          R   5</span></a>
<a class="sourceLine" id="cb26-10" data-line-number="10"></a>
<a class="sourceLine" id="cb26-11" data-line-number="11"><span class="co"># Now let&#39;s calcualte the PCA for all the data (not just the</span></a>
<a class="sourceLine" id="cb26-12" data-line-number="12"><span class="co"># top variant genes)</span></a>
<a class="sourceLine" id="cb26-13" data-line-number="13">pca_all &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(<span class="kw">log2</span>(cpm <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)))</a>
<a class="sourceLine" id="cb26-14" data-line-number="14"></a>
<a class="sourceLine" id="cb26-15" data-line-number="15"><span class="co"># Let&#39;s make a temporary demographic without the species</span></a>
<a class="sourceLine" id="cb26-16" data-line-number="16">tmp_demo &lt;-<span class="st"> </span>demo[<span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>)]</a>
<a class="sourceLine" id="cb26-17" data-line-number="17"></a>
<a class="sourceLine" id="cb26-18" data-line-number="18"><span class="co"># Let&#39;s include the first PC into the demo file.  This is the</span></a>
<a class="sourceLine" id="cb26-19" data-line-number="19"><span class="co"># one that explain most of the variance in the data.</span></a>
<a class="sourceLine" id="cb26-20" data-line-number="20">tmp_demo<span class="op">$</span>pca1 &lt;-<span class="st"> </span>pca_all<span class="op">$</span>x[, <span class="dv">1</span>]</a>
<a class="sourceLine" id="cb26-21" data-line-number="21"></a>
<a class="sourceLine" id="cb26-22" data-line-number="22"><span class="co"># Make a model matrix for the model</span></a>
<a class="sourceLine" id="cb26-23" data-line-number="23">mm &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">model.matrix</span>(<span class="op">~</span>., tmp_demo))</a>
<a class="sourceLine" id="cb26-24" data-line-number="24"></a>
<a class="sourceLine" id="cb26-25" data-line-number="25"><span class="co"># Now we are going to fit a moodel between the first</span></a>
<a class="sourceLine" id="cb26-26" data-line-number="26"><span class="co"># component and the other covaraites</span></a>
<a class="sourceLine" id="cb26-27" data-line-number="27">fit1 =<span class="st"> </span><span class="kw">lm</span>(pca1 <span class="op">~</span><span class="st"> </span>., <span class="dt">data =</span> mm)</a>
<a class="sourceLine" id="cb26-28" data-line-number="28"></a>
<a class="sourceLine" id="cb26-29" data-line-number="29"><span class="co"># Let&#39;s transform the fit into a dataframe with</span></a>
<a class="sourceLine" id="cb26-30" data-line-number="30"><span class="co"># library(broom)</span></a>
<a class="sourceLine" id="cb26-31" data-line-number="31">df &lt;-<span class="st"> </span><span class="kw">tidy</span>(fit1)[<span class="op">-</span><span class="dv">1</span>, ]</a>
<a class="sourceLine" id="cb26-32" data-line-number="32"></a>
<a class="sourceLine" id="cb26-33" data-line-number="33"><span class="co"># Let&#39;s check the info reported</span></a>
<a class="sourceLine" id="cb26-34" data-line-number="34"><span class="kw">head</span>(df)</a>
<a class="sourceLine" id="cb26-35" data-line-number="35"><span class="co">## # A tibble: 5 x 5</span></a>
<a class="sourceLine" id="cb26-36" data-line-number="36"><span class="co">##   term        estimate std.error statistic p.value</span></a>
<a class="sourceLine" id="cb26-37" data-line-number="37"><span class="co">##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;</span></a>
<a class="sourceLine" id="cb26-38" data-line-number="38"><span class="co">## 1 SexM          -42.8      24.3     -1.76   0.0904</span></a>
<a class="sourceLine" id="cb26-39" data-line-number="39"><span class="co">## 2 Age             1.38      1.45     0.949  0.352 </span></a>
<a class="sourceLine" id="cb26-40" data-line-number="40"><span class="co">## 3 RIN             1.95      8.40     0.233  0.818 </span></a>
<a class="sourceLine" id="cb26-41" data-line-number="41"><span class="co">## 4 HemisphereR    18.3      21.3      0.858  0.400 </span></a>
<a class="sourceLine" id="cb26-42" data-line-number="42"><span class="co">## 5 PMI            -2.40      2.35    -1.02   0.317</span></a>
<a class="sourceLine" id="cb26-43" data-line-number="43"></a>
<a class="sourceLine" id="cb26-44" data-line-number="44"><span class="co"># The Pvalue reflect how much the covariates are associated</span></a>
<a class="sourceLine" id="cb26-45" data-line-number="45"><span class="co"># with the first component.  Transform the p-value into a</span></a>
<a class="sourceLine" id="cb26-46" data-line-number="46"><span class="co"># log10 scale.</span></a>
<a class="sourceLine" id="cb26-47" data-line-number="47">df<span class="op">$</span>log10 =<span class="st"> </span><span class="op">-</span><span class="kw">log10</span>(df<span class="op">$</span>p.value)</a>
<a class="sourceLine" id="cb26-48" data-line-number="48"></a>
<a class="sourceLine" id="cb26-49" data-line-number="49"><span class="co"># Let&#39;s visualize the association based on this modeling</span></a>
<a class="sourceLine" id="cb26-50" data-line-number="50"><span class="kw">ggbarplot</span>(df, <span class="dt">x =</span> <span class="st">&quot;term&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;log10&quot;</span>, <span class="dt">fill =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">color =</span> <span class="st">&quot;blue&quot;</span>, </a>
<a class="sourceLine" id="cb26-51" data-line-number="51">  <span class="dt">x.text.angle =</span> <span class="dv">90</span>, <span class="dt">ylab =</span> <span class="st">&quot;-log10(P)&quot;</span>, <span class="dt">xlab =</span> <span class="st">&quot;Covariates&quot;</span>, </a>
<a class="sourceLine" id="cb26-52" data-line-number="52">  <span class="dt">rotate =</span> <span class="ot">TRUE</span>, <span class="dt">ggtheme =</span> <span class="kw">theme_classic</span>()) <span class="op">+</span><span class="st"> </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="fl">1.3</span>, </a>
<a class="sourceLine" id="cb26-53" data-line-number="53">  <span class="dt">linetype =</span> <span class="st">&quot;dotted&quot;</span>, <span class="dt">color =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">size =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylim</span>(<span class="dv">0</span>, <span class="dv">5</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb26-54" data-line-number="54"><span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;PCA1 vs Covariates&quot;</span>)</a></code></pre></div>
<p><img src="bookdown-demo_files/figure-html/covaraites-1.png" width="65%" style="display: block; margin: auto;" /></p>
</div>
<div id="save-the-data" class="section level3">
<h3><span class="header-section-number">2.0.6</span> Save the data</h3>
<p>Now it’s time to save all the data we generated.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"></a>
<a class="sourceLine" id="cb27-2" data-line-number="2"><span class="kw">save</span>(exp, cpm, rpkm, tpm, quantGC, demo, <span class="dt">file =</span> <span class="st">&quot;peb_data/Normalized_data.RData&quot;</span>)</a></code></pre></div>
</div>
<div id="exercise-for-data-exploration-chapter" class="section level3">
<h3><span class="header-section-number">2.0.7</span> Exercise for Data Exploration chapter</h3>
<ul>
<li><p>Do data exploration for RPKM.</p></li>
<li><p>Do data exploration for TPM</p></li>
<li><p>Do data exploration for quantGC.</p></li>
<li><p>Covaraite association with different normalized values.</p></li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introPipes.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="DiffExpAnalysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook/js/app.min.js"></script>
<script src="libs/gitbook/js/lunr.js"></script>
<script src="libs/gitbook/js/plugin-search.js"></script>
<script src="libs/gitbook/js/plugin-sharing.js"></script>
<script src="libs/gitbook/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook/js/plugin-bookdown.js"></script>
<script src="libs/gitbook/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/02-DataExploration.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
